### PR TITLE
Update and cleanup MediaSessions

### PR DESCRIPTION
**Pull request by MrTimscampi**
**Changes**

* Updates the MediaSessions feature to the current spec, removing non-existing parameters.
* Send the full artwork image to NativeShell

**Issues**

### CODE REVIEW

File: src/components/playback/mediasession.js [hunk 1]
```diff
@@ -1,45 +1,26 @@
...
-    function seriesImageUrl(item, options) {
+    let currentPlayer;
 
+    function seriesImageUrl(item, options = {}, type = options.type || 'Primary') {
         if (item.Type !== 'Episode') {
             return null;
-        }
-
-        options = options || {};
-        options.type = options.type || "Primary";
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:21]: this setting is missed, what if server or api client uses that?..

File: src/components/playback/mediasession.js [hunk 2]
```diff
@@ -49,121 +30,96 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
         return null;
     }
 
-    function imageUrl(item, options) {
-
-        options = options || {};
-        options.type = options.type || "Primary";
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:55]: same, what if `options.type` is unset? before it defaulted it to `Primary` but now it would remain unset

File: src/components/playback/mediasession.js [hunk 3]
```diff
@@ -49,121 +30,96 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
...
+
             return connectionManager.getApiClient(item.ServerId).getScaledImageUrl(item.AlbumId, options);
         }
 
         return null;
     }
 
-    function pushImageUrl(item, imageOptions, list) {
-        var url = seriesImageUrl(item, imageOptions) || imageUrl(item, imageOptions);
+    function pushImageUrl(item, imageOptions = {}) {
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:47]: the name now looks wrong

File: src/components/playback/mediasession.js [hunk 4]
```diff
@@ -49,121 +30,96 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
...
 
-        var playState = state.PlayState || {};
-        var parts = nowPlayingHelper.getNowPlayingNames(item);
-        var artist = parts[parts.length - 1].text;
-        var title = parts.length === 1 ? '' : parts[0].text;
-        var albumArtist;
-
-        if (item.AlbumArtists && item.AlbumArtists[0]) {
-            albumArtist = item.AlbumArtists[0].Name;
-        }
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:134]: so this is now lost?

File: src/components/playback/mediasession.js [hunk 5]
```diff
@@ -49,121 +30,96 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
...
                 title: title,
                 artist: artist,
                 album: album,
-                artwork: getImageUrls(item),
-                albumArtist: albumArtist,
-                currentTime: currentTime,
-                duration: duration,
-                paused: isPaused,
-                itemId: itemId,
-                mediaType: item.MediaType
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:157]: why is this removed?

File: src/components/playback/mediasession.js [hunk 6]
```diff
@@ -49,121 +30,96 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
...
             });
         } else {
-            var imageUrl = [];
-            pushImageUrl(item, {maxHeight: 400}, imageUrl);
-
-            if (imageUrl.length) {
-                imageUrl = imageUrl[0].src;
-            } else {
-                imageUrl = null;
+            let itemImageUrl = seriesImageUrl(item) || imageUrl(item);
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:120]: why `maxHeight` is gone?

File: src/components/playback/mediasession.js [hunk 7]
```diff
@@ -49,121 +30,96 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
...
+            if (!itemImageUrl) {
+                itemImageUrl = null;
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:122]: cannot you do stuff like
```js
let itemImageUrl = seriesImageUrl(item) || imageUrl(item) || null;
```
and remove this `if`?

File: src/components/playback/mediasession.js [hunk 8]
```diff
@@ -228,8 +172,6 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
     }
 
     function hideMediaControls() {
-        lastUpdateTime = 0;
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:231]: hmm, you sure?..

File: src/components/playback/mediasession.js [hunk 9]
```diff
@@ -261,8 +202,8 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
     function execute(name) {
         playbackManager[name](currentPlayer);
     }
-    if (navigator.mediaSession) {
 
+    if (navigator.mediaSession) {
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:206]: why're you checking it differently? in code above you used `if 'mediaSession' in navigator`

File: src/components/playback/mediasession.js [hunk 10]
```diff
@@ -1,45 +1,26 @@
...
-
-                return connectionManager.getApiClient(item.ServerId).getScaledImageUrl(item.SeriesId, options);
-            }
-        }
-
-        if (options.type === 'Thumb') {
+        } else if (type === 'Primary' && item.SeriesPrimaryImageTag) {
+            options.tag = item.SeriesPrimaryImageTag;
 
+            return connectionManager.getApiClient(item.ServerId).getScaledImageUrl(item.SeriesId, options);
```
Comment (ferferga) [inline src/components/playback/mediasession.js:17]: This line breaks

File: src/components/playback/mediasession.js (hunk 1 above)
Comment (ferferga) [reply src/components/playback/mediasession.js:21]: @MrTimscampi I think this might be related to my review

File: src/components/playback/mediasession.js [hunk 11]
```diff
@@ -1,45 +1,26 @@
...
     // Reports media playback to the device for lock screen control
 
-    var currentPlayer;
-    var lastUpdateTime = 0;
-
-    function seriesImageUrl(item, options) {
+    let currentPlayer;
 
+    function seriesImageUrl(item, options = {}, type = options.type || 'Primary') {
         if (item.Type !== 'Episode') {
```
Comment (ferferga) [inline src/components/playback/mediasession.js:12]:
```suggestion
        options = options || {};
        options.type = options.type || "Primary";
        if (item.Type !== 'Episode') {
```

Adding this back fixes what I commented in my review, as pointed out by @JustAMan.

Comment (MrTimscampi) [reply src/components/playback/mediasession.js:12]: The default for "options" is set in line 11.

The optional "type" argument, which is new, default to either the value of options.type or "Primary", as defaults for arguments have access to all of the previous argument's values.

Comment (MrTimscampi) [reply src/components/playback/mediasession.js:12]: Oh, I see what you mean.

File: src/components/playback/mediasession.js (hunk 4 above)
Comment (MrTimscampi) [reply src/components/playback/mediasession.js:134]: It never worked, since it's only used in the MediaSessions object that is passed to the browser and, afaik, that has never supported album artists or some of the other properties set before...

File: src/components/playback/mediasession.js (hunk 5 above)
Comment (MrTimscampi) [reply src/components/playback/mediasession.js:157]: Because the MediaMetadata object never supported them in the first place.

See here: https://developer.mozilla.org/en-US/docs/Web/API/MediaMetadata/MediaMetadata

File: src/components/playback/mediasession.js (hunk 6 above)
Comment (MrTimscampi) [reply src/components/playback/mediasession.js:124]: This PR originated from @dkanada noticing that artwork on Android was VERY low resolution, most notably the album art on the lockscreen.

Since I couldn't find any recommendation from Google on which resolution to use, I removed the limit.

File: src/components/playback/mediasession.js (hunk 7 above)
Comment (MrTimscampi) [reply src/components/playback/mediasession.js:126]: I'm not sure if it would work. And in any case, it'd probably make more sense to return `null` from seriesImageUrl and imageUrl.

File: src/components/playback/mediasession.js (hunk 8 above)
Comment (MrTimscampi) [reply src/components/playback/mediasession.js:231]: It's unused, according to ESLint...

File: src/components/playback/mediasession.js (hunk 9 above)
Comment (MrTimscampi) [reply src/components/playback/mediasession.js:210]: This was an original line that I moved down. I'll change the check too, for consistency.

File: src/components/playback/mediasession.js (hunk 6 above)
Comment (ferferga) [reply src/components/playback/mediasession.js:126]: I think it makes sense to leave it at full res, so each device decides better how to fit it. It's only one image as well.

File: src/components/playback/mediasession.js [hunk 12]
```diff
@@ -54,14 +58,16 @@ import connectionManager from 'connectionManager';
                 src: url,
                 sizes: height + 'x' + height
             };
+        } else {
+            return null;
```
Comment (JustAMan) [inline src/components/playback/mediasession.js:62]: won't this break later on when the result of `getImageUrls()` containing `null`s would be used by something else?

File: src/components/playback/mediasession.js (hunk 6 above)
Comment (JustAMan) [reply src/components/playback/mediasession.js:126]: Maybe we should limit it to FullHD or smth? I'm imagining the 8K artwork being transferred to the phone over not-so-fast 3G...

File: src/components/playback/mediasession.js (hunk 12 above)
Comment (MrTimscampi) [reply src/components/playback/mediasession.js:62]: It would already break since nothing is currently returned.

I added a explicit check to omit pushing the image to the array if it's null.

File: src/components/playback/mediasession.js (hunk 6 above)
Comment (MrTimscampi) [reply src/components/playback/mediasession.js:129]: Apparently, the recommended size for Spotify is 3000x3000.

However, there is not resolution limit, the only real limit being a size under 4MB.

I think limiting the max height to 3000px should work well enough.

**Comment by MrTimscampi** [comment]
@JustAMan Can't respond to your comment on line 120, but I looked a bit into it and the recommendation for Spotify seems to be 3000x3000.

We could set it to a maxHeight of 3000px.

**Comment by JustAMan** [comment]
@MrTimscampi can you do so then?
@ferferga can you re-review?..

File: src/components/playback/mediasession.js [hunk 13]
```diff
@@ -1,45 +1,28 @@
-define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], function (playbackManager, nowPlayingHelper, events, connectionManager) {
-    "use strict";
-
-    // no support for mediaSession
-    if (!navigator.mediaSession && !window.NativeShell) {
```
Comment (ferferga) [inline src/components/playback/mediasession.js:5]: Why this is removed?

File: src/components/playback/mediasession.js [hunk 14]
```diff
@@ -49,122 +32,101 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
...
+        const duration = parseInt(item.RunTimeTicks ? (item.RunTimeTicks / 10000) : 0);
+        const currentTime = parseInt(playState.PositionTicks ? (playState.PositionTicks / 10000) : 0);
 
-        var isPaused = playState.IsPaused || false;
-        var canSeek = playState.CanSeek || false;
+        const isPaused = playState.IsPaused || false;
+        const canSeek = playState.CanSeek || false;
 
-        if (navigator.mediaSession) {
+        if ('mediaSession' in navigator) {
```
Comment (ferferga) [inline src/components/playback/mediasession.js:121]: I would remove all the checking here and just leave the one done at the beginning

File: src/components/playback/mediasession.js [hunk 15]
```diff
@@ -228,17 +178,14 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
     }
 
     function hideMediaControls() {
-        lastUpdateTime = 0;
-
-        if (navigator.mediaSession) {
+        if ('mediaSession' in navigator) {
```
Comment (ferferga) [inline src/components/playback/mediasession.js:181]: Same here

File: src/components/playback/mediasession.js [hunk 16]
```diff
@@ -261,8 +208,8 @@ define(['playbackManager', 'nowPlayingHelper', 'events', 'connectionManager'], f
     function execute(name) {
         playbackManager[name](currentPlayer);
     }
-    if (navigator.mediaSession) {
 
+    if ('mediaSession' in navigator) {
```
Comment (ferferga) [inline src/components/playback/mediasession.js:212]: same

File: src/components/playback/mediasession.js (hunk 14 above)
Comment (MrTimscampi) [reply src/components/playback/mediasession.js:121]: They have different purposes.

The one at the beginning is to check if either the Media Session API or Native Shell is available.  
This one is specifically to check if the Media Session API is available, since a client without Media Session but with Native Shell will load this file.

**Comment by MrTimscampi** [comment]
@ferferga Since you can't return from an ES6 module, I moved the original check to the only loading point I could find.

As for the other check, as explained, they're there in case the client has NativeShell but doesn't support the Media Session API.

**Comment by ferferga** [comment]
@MrTimscampi This is exactly what I was thinking!

Checking out this branch to (hopefully now) say that it's good to go! :)
